[SwiftUI] Introduce WebPage

rr-codes opened the pull request:
#### e0e599e62b7ed43af3f348430aef948f2562ca5e
<pre>
[SwiftUI] Introduce WebPage
<a href="https://bugs.webkit.org/show_bug.cgi?id=283244">https://bugs.webkit.org/show_bug.cgi?id=283244</a>
<a href="https://rdar.apple.com/140050033">rdar://140050033</a>

Reviewed by Wenson Hsieh.

Introduce a very bare-bones `WebPage` type. Currently, it can only load url requests or html strings
and track the corresponding navigation events.

* Source/WebKit/UIProcess/API/Swift/WKNavigationDelegateAdapter.swift: Added.
(WKNavigationDelegateAdapter.navigationProgressContinuation):
(WKNavigationDelegateAdapter.webView(_:didStartProvisionalNavigation:)):
(WKNavigationDelegateAdapter.webView(_:didReceiveServerRedirectForProvisionalNavigation:)):
(WKNavigationDelegateAdapter.webView(_:didCommit:)):
(WKNavigationDelegateAdapter.webView(_:didFinish:)):
(WKNavigationDelegateAdapter.webView(_:didFailProvisionalNavigation:withError:)):
(WKNavigationDelegateAdapter.webView(_:didFail:withError:)):
* Source/WebKit/UIProcess/API/Swift/WebPage+Navigation.swift: Added.
(Navigation.makeAsyncIterator):
(Iterator.source):
(Iterator.next):
* Source/WebKit/UIProcess/API/Swift/WebPage.swift: Added.
(WebPage.navigate(to:)):
(WebPage.navigate(to:baseURL:)):
(WebPage.createNavigation(_:)):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit Swift/WebPageNavigation.swift: Added.
(URL.aboutBlank):
(WebPageNavigationTests.basicNavigation):
(WebPageNavigationTests.failedProvisionNavigationThrowsError):

Canonical link: <a href="https://commits.webkit.org/286886@main">https://commits.webkit.org/286886@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/50d75a490f6f7cb630f7c1115ee7ed5467c053fd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77364 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56399 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30280 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81940 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28646 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79481 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65547 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4696 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60614 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18644 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80431 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50593 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66409 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40912 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47995 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23907 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-masking/clip-path/animations/clip-path-animation-font-size-mixed-change.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26969 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/69106 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24245 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83353 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4744 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3227 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68878 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4900 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66379 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68136 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12135 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10237 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11984 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4691 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/7506 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4710 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/8145 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6469 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Tools/TestWebKitAPI/Tests/WebKit Swift/WebPageNavigation.swift:
```diff
@@ -0,0 +1,83 @@
+        
+        await #expect(throws: WebPage.NavigationError.self) {
+            for try await _ in page.navigate(to: request) {
+            }
```
Hm... this is an interesting API shape!

I'm curious what happens in the cases where:

- The web content process is terminated during navigation (does that invoke `-didFailNavigation`? or should the Swift equivalent of `-webViewWebContentProcessDidTerminate:` be treated as a "terminal" state, similar to `-didFinishNavigation`). It might be good to have a test that exercises this..
- How would a SwiftUI WebKit client respond correctly to navigation that's either programmatically triggered by the page or triggered by the user? (e.g. the client wants to be informed when a load is committed so that it can show an indeterminate progress indicator in the UI)

r=me, assuming we can figure out how to keep the open source build passing.

axiixc commented on Source/WebKit/UIProcess/API/Swift/WKNavigationDelegateAdapter.swift:
```diff
@@ -0,0 +1,60 @@
+
+import Foundation
+
+@objc(WKNavigationDelegateAdapter)
```
You don't really need this line unless you're going to refer to this class name by string. By virtue of inheriting `NSObject` it is already implicitly `@objc`, and relevant `WKNavigationDelegate` methods are likewise implicitly `@objc`.

rr-codes replied:
I originally tried without it, but the build failed because it's an "inappropriate Obj C name" because the mangled name doesn't start with WK

axiixc replied:
Ah, I'd probably suggest we just add the Swift mangled module name prefix to that allow list instead. There's no real danger here of namespace collisions.
```
_TtC12WebKit _TtCC12WebKit
```

---

Having said that, don't forget that `@objc()` isn't appropriate for publishing ObjC classes implemented in Swift as API (or really in Private headers either), it's basically only okay within a module. For that you'll want to use `@objc @implementation extension` and keep writing real headers.

This class is `internal` so that doesn't apply here.

axiixc commented on Source/WebKit/UIProcess/API/Swift/WKNavigationDelegateAdapter.swift:
```diff
@@ -0,0 +1,60 @@
+    // MARK: Navigation progress reporting
+    
+    func webView(_ webView: WKWebView, didStartProvisionalNavigation navigation: WKNavigation!) {
+        self.navigationProgressContinuation?.yield(.startedProvisionalNavigation)
```
Nit: We've mostly found that using `self.` explicitly is difficult since it is optional, so most Swift code has dropped this.

rr-codes replied:
will fix!

axiixc commented on Source/WebKit/UIProcess/API/Swift/WKNavigationDelegateAdapter.swift:
```diff
@@ -0,0 +1,60 @@
+
+    func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
+        self.navigationProgressContinuation?.yield(.finished)
+        self.navigationProgressContinuation?.finish()
```
Should it also get assigned to `nil` to prevent further use? What ensures this doesn't get used after being finished?

rr-codes replied:
it gets assigned to `nil` in the `continuation.onTermination ` block in WebPage

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage+Navigation.swift:
```diff
@@ -0,0 +1,85 @@
+
+import Foundation
+
+@_spi(Internal)
```
Curious why we're using SPI for this?

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage+Navigation.swift:
```diff
@@ -0,0 +1,85 @@
+        case failed(underlyingError: any Error)
+    }
+    
+    struct Navigation: AsyncSequence, Sendable {
```
I need to double check on naming guidance, but I think this would be `Navigations` since it's a sequence. In any case `Navigation` seems like a bad name to burn for this since I'd expect that type to describe an individual navigation.

rr-codes replied:
it is one individual navigation; just with a stream of events marking the navigation's progress

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage+Navigation.swift:
```diff
@@ -0,0 +1,85 @@
+        
+        private let source: AsyncThrowingStream<Self.Element, any Error>
+        
+        init(source: AsyncThrowingStream<Self.Element, any Error>) {
```
I don't think we've settled on this for coding style, but I've been pushing for `typealias` > `static` > `init` > `var`s > methods > marked segments as the high level type definition order.

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage+Navigation.swift:
```diff
@@ -0,0 +1,85 @@
+            do {
+                return try await self.source.next()
+            } catch {
+                throw error as! Failure
```
Why do you need the cast here? I think this is because you're opting in to typed throws, which I'm not sure we really need to do.

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage.swift:
```diff
@@ -0,0 +1,81 @@
+
+@_spi(Internal)
+@MainActor public class WebPage {
+    // MARK: Non-public properties.
```
```suggestion
    // MARK: Non-public Properties
```
Kinda a weird mark name though. Should we organize these functionally rather than by access level?

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage.swift:
```diff
@@ -0,0 +1,81 @@
+    // MARK: Public initializers
+    
+    public init() {
+        self.backingWebView = WKWebView(frame: .zero)
```
This gets in to a bigger discussion, but we might want to plan to not eagerly instantiate the web view. This could be a `lazy` property that gets set up only as needed to keep the cost of this type down.

axiixc commented on Tools/TestWebKitAPI/Tests/WebKit Swift/WebPageNavigation.swift:
```diff
@@ -0,0 +1,93 @@
+
+// MARK: Helper extension functions
+
+fileprivate extension Array {
```
`fileprivate` is just `private` at top level like this. I will also warn that adding access control to extensions can bite you. Sometimes people will stick `public` on there for some API and then later add what they think of as `internal` helpers which are actually made `public` (though that's less of a danger for more restrictive levels like this).

emw-apple commented on Tools/TestWebKitAPI/Tests/WebKit Swift/WebPageNavigation.swift:
```diff
@@ -0,0 +1,93 @@
+ // ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ // THE POSSIBILITY OF SUCH DAMAGE.
+
+#if ENABLE_SWIFTUI && canImport(Testing) && compiler(>=6.0)
```
Do we build a configuration where the compiler is Swift 6 but we cannot import Swift Testing?

emw-apple commented on Source/WebKit/Configurations/Base.xcconfig:
```diff
@@ -120,6 +120,17 @@ SWIFT_VERSION = 5.0;
 OTHER_SWIFT_FLAGS = $(inherited) $(OTHER_SWIFT_FLAGS$(WK_MACOS_1400)) -no-warnings-as-errors -Xfrontend -experimental-spi-only-imports -Xfrontend -enable-experimental-concurrency -Xfrontend -enable-upcoming-feature -Xfrontend IsolatedDefaultValues $(OTHER_SWIFT_FLAGS_AVAILABILITY_$(USE_INTERNAL_SDK));
 OTHER_SWIFT_FLAGS_AVAILABILITY_YES = @$(WK_WEBKITADDITIONS_HEADERS_FOLDER_PATH)/Scripts/availability-definitions.txt;
 
+SWIFT_ACTIVE_COMPILATION_CONDITIONS = $(inherited) $(SWIFT_ACTIVE_COMPILATION_CONDITIONS_$(WK_PLATFORM_NAME));
+SWIFT_ACTIVE_COMPILATION_CONDITIONS_iphoneos = $(SWIFT_ACTIVE_COMPILATION_CONDITIONS$(WK_IOS_18));
+SWIFT_ACTIVE_COMPILATION_CONDITIONS_iphonesimulator = $(SWIFT_ACTIVE_COMPILATION_CONDITIONS$(WK_IOS_18));
+SWIFT_ACTIVE_COMPILATION_CONDITIONS_xros = $(SWIFT_ACTIVE_COMPILATION_CONDITIONS$(WK_IOS_18));
+SWIFT_ACTIVE_COMPILATION_CONDITIONS_xrsimulator = $(SWIFT_ACTIVE_COMPILATION_CONDITIONS$(WK_IOS_18));
+SWIFT_ACTIVE_COMPILATION_CONDITIONS_IOS_SINCE_18 = ENABLE_SWIFTUI;
+SWIFT_ACTIVE_COMPILATION_CONDITIONS_maccatalyst = $(SWIFT_ACTIVE_COMPILATION_CONDITIONS_maccatalyst$(WK_MACOS_1500));
+SWIFT_ACTIVE_COMPILATION_CONDITIONS_maccatalyst_MACOS_SINCE_1500 = ENABLE_SWIFTUI;
+SWIFT_ACTIVE_COMPILATION_CONDITIONS_macosx = $(SWIFT_ACTIVE_COMPILATION_CONDITIONS$(WK_MACOS_1500));
+SWIFT_ACTIVE_COMPILATION_CONDITIONS_MACOS_SINCE_1500 = ENABLE_SWIFTUI;
```
I think you can safely check `$(WK_MACOS_1500)` and `$(WK_IOS_18)` without checking WK_PLATFORM_NAME first. Unless you found otherwise?

rr-codes replied:
I need it to not be true on watchOS or visionOS, which is what happened when I stuck with just the Mac and iOS pieces

rr-codes committed:
[SwiftUI] Introduce WebPage
https://bugs.webkit.org/show_bug.cgi?id=283244
rdar://140050033

Reviewed by Wenson Hsieh.

Introduce a very bare-bones `WebPage` type. Currently, it can only load url requests or html strings
and track the corresponding navigation events.

* Source/WebKit/UIProcess/API/Swift/WKNavigationDelegateAdapter.swift: Added.
(WKNavigationDelegateAdapter.navigationProgressContinuation):
(WKNavigationDelegateAdapter.webView(_:didStartProvisionalNavigation:)):
(WKNavigationDelegateAdapter.webView(_:didReceiveServerRedirectForProvisionalNavigation:)):
(WKNavigationDelegateAdapter.webView(_:didCommit:)):
(WKNavigationDelegateAdapter.webView(_:didFinish:)):
(WKNavigationDelegateAdapter.webView(_:didFailProvisionalNavigation:withError:)):
(WKNavigationDelegateAdapter.webView(_:didFail:withError:)):
* Source/WebKit/UIProcess/API/Swift/WebPage+Navigation.swift: Added.
(Navigation.makeAsyncIterator):
(Iterator.source):
(Iterator.next):
* Source/WebKit/UIProcess/API/Swift/WebPage.swift: Added.
(WebPage.navigate(to:)):
(WebPage.navigate(to:baseURL:)):
(WebPage.createNavigation(_:)):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit Swift/WebPageNavigation.swift: Added.
(URL.aboutBlank):
(WebPageNavigationTests.basicNavigation):
(WebPageNavigationTests.failedProvisionNavigationThrowsError):

Canonical link: https://commits.webkit.org/286886@main
